airflow/providers/amazon/aws/utils/redshift.py

```diff
@@ -0,0 +1,49 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import logging
+
+from botocore.credentials import ReadOnlyCredentials
+
+log = logging.getLogger(__name__)
+
+
+def build_credentials_block(credentials: ReadOnlyCredentials) -> str:
+    """
+    Generate AWS credentials block for Redshift COPY and UNLOAD
+    commands, as noted in AWS docs
+    https://docs.aws.amazon.com/redshift/latest/dg/copy-parameters-authorization.html#copy-credentials
+
+    :param credentials: ReadOnlyCredentials object from `botocore`
+    :return: str
+    """
+    if credentials.token and credentials.access_key.startswith("ASIA"):
```

Comment:
What's this for?

Reply:
When a principal assumes an IAM role, Security Token Service (STS) hands out a set of values: access key, secret key and a token. These are temporary credentials valid for one hour by default. They are easily recognized as the access key is starting with `ASIA` which is different from regular credentials that are starting with `AKIA`

If a principal wishes to use these credentials then the token must be included along with the access key and secret key. The token does not exist when using regular credentials starting with `AKIA`.

Issue that we encountered when using Airflow is exactly that: Airflow runs under an IAM role but assumes another IAM role to get access to S3 in order to perform the `UNLOAD` command. However, since the token is not included the authentication does not work and role assumption fails, and so does the `UNLOAD` command.

Reply:
Sorry, I should have been more clear:

Why do we care what the prefix is -- why isn't this `if credentials.token:` -- i.e. if there is a token, use it.